source/JsMaterialX/test/customBindings.spec.js

```diff
@@ -0,0 +1,225 @@
+        const token = new mx.Token(node, 'token');
+        expect(token.validate()).to.be.true;
+        token.setUnitType('bogus');
+        expect(token.validate()).to.be.false;
+        expect(token.validate(message)).to.be.false;
+        expect(message.message).to.include('Unit type definition does not exist in document')
+    });
+
+    it('StringResolver name substitution getters', () => {
+        const fnTestData = {
+            fnKey: 'fnValue',
+            fnKey1: 'fnValue1'
+        };
+        const fnTestKeys = Object.keys(fnTestData);
+
+        const gnTestData = {
+            gnKey: 'gnValue',
+            gnKey1: 'gnValue1'
+        };
+        const gnTestKeys = Object.keys(gnTestData);
+
+        const resolver = mx.StringResolver.create();
+
+        resolver.setFilenameSubstitution(fnTestKeys[0], fnTestData[fnTestKeys[0]]);
+        resolver.setFilenameSubstitution(fnTestKeys[1], fnTestData[fnTestKeys[1]]);
+        const fnSubs = resolver.getFilenameSubstitutions();
+        expect(fnSubs).to.be.instanceof(Object);
+        expect(Object.keys(fnSubs).length).to.equal(2);
+        expect(fnSubs).to.deep.equal(fnTestData);
+
+        resolver.setGeomNameSubstitution(gnTestKeys[0], gnTestData[gnTestKeys[0]]);
+        resolver.setGeomNameSubstitution(gnTestKeys[1], gnTestData[gnTestKeys[1]]);
+        const gnSubs = resolver.getGeomNameSubstitutions();
+        expect(gnSubs).to.be.instanceof(Object);
+        expect(Object.keys(gnSubs).length).to.equal(2);
+        expect(gnSubs).to.deep.equal(gnTestData);
+    });
+
+    it('getShaderNodes', () => {
+        const doc = mx.createDocument();
+        const mtlxStrs = getMtlxStrings(['MaterialBasic.mtlx'], '../../../resources/Materials/Examples/Syntax');
+        mx.readFromXmlString(doc, mtlxStrs[0]);
+        let matNodes = doc.getMaterialNodes();
+        expect(matNodes.length).to.equal(2);
+        const matNode = matNodes[0];
+
+        // Should return a surface shader node when called without optional parameters
+        let shaderNodes = mx.getShaderNodes(matNode);
+        expect(shaderNodes).to.be.instanceof(Array);
+        expect(shaderNodes.length).to.equal(1);
+        expect(shaderNodes[0].getType()).to.equal(mx.SURFACE_SHADER_TYPE_STRING);
+
+        // Should return a shader node of the given optional type
+        shaderNodes = mx.getShaderNodes(matNode, mx.DISPLACEMENT_SHADER_TYPE_STRING);
+        expect(shaderNodes).to.be.instanceof(Array);
+        expect(shaderNodes.length).to.equal(1);
+        expect(shaderNodes[0].getType()).to.equal(mx.DISPLACEMENT_SHADER_TYPE_STRING);
+        shaderNodes = mx.getShaderNodes(matNode, 'bogus');
+        expect(shaderNodes).to.be.instanceof(Array);
+        expect(shaderNodes.length).to.equal(0);
+
+        // TODO: Couldn't find a case / sample file where an input node has a target definition
+        // Should filter shader nodes based on the given optional target type
```

Comment:
Cannot you create them in place? like `doc.addTargetDef()`

Reply:
Hm, not sure if this is what the flag is related to, but I'll give it a try! 👍